feat: add post on We Love Speed 2023 (#426)
BedrockStreaming/tech.bedrockstreaming.com: _posts/2023-06-08-we-love-speed-2023.md

### Priorization

### Community

Maxime ?

## Avec quels moyens s’attaquer à la performance ?

When analyzing a website's performance, it's essential to take several key criteria into account. Here are some of the commonly used performance analysis criteria:

- Largest Contentful Paint (LCP): LCP measures the time it takes for the largest visible element on the page (such as an image or block of text) to be completely rendered on screen. A fast LCP indicates fast page loading and a better user experience.

- Cumulative Layout Shift (CLS): CLS measures the visual stability of the page by evaluating changes in element position during loading. A low CLS means that page elements do not move unexpectedly, providing a more stable and pleasant user experience.

- First Contentful Paint (FCP): FCP measures the time elapsed between the start of navigation and the display of the first element on the screen, whether text, image or other visual element. A fast FCP indicates that the page is loading quickly and gives the impression of responsiveness.

> 
> 
> - Time to Interactive (TTI): TTI measures the time it takes for the page to become fully interactive, i.e. users can interact with elements on the page, click on links, fill in forms, etc. A fast TTI is crucial to the success of a site. A fast TTI is crucial for user engagement.
> 
> 
> - First Input Delay (FID): FID measures the delay between the user's first interaction with the page (for example, a click on a button) and the page's actual response to that interaction. A short FID indicates high responsiveness and a smooth user experience.
> 
> 
> - Interaction to Next Paint (INP): INP is a performance metric that measures the delay between a user interaction and the moment when the next visual update is rendered on the screen. It is a specific metric that evaluates the responsiveness of the website to user actions. A short INP indicates that the website reacts quickly to user actions, resulting in a pleasant, lag-free user experience.
> 
> To measure/evaluate your site's performance according to these criteria, there are 3 types of sources: Local (devtools), Synthetic (CI/CD / online scan tools), RUM (Real User Monitoring).
> 
> - Local development tools - devtools :
>   - Chrome DevTools: A set of development tools integrated into the Chrome browser, for analyzing performance, debugging JavaScript code, inspecting DOM elements, etc.
>   - Firefox DevTools: Tools similar to Chrome DevTools, but specific to the Firefox browser. They offer debugging, performance auditing and inspection of web elements.
>   - Lighthouse DevTools (Chrome): A Chrome DevTools extension that provides automated audits to evaluate performance, accessibility, SEO optimization, etc.
> 
> - Synthetic testing - CI/CD and online analysis tools :
>   - WebPageTest: An online tool for evaluating website performance by running tests on different browser configurations, network connections and geographical locations.
>   - Lighthouse CI: A Lighthouse-based continuous integration (CI) solution that automates web performance and quality audits with every deployment.
> 
> - Real User Monitoring (RUM):
>   - New Relic: An application performance monitoring platform that provides real-time visibility into the performance of your web applications, including usage metrics, load times, etc.
>   - Grafana: A data visualization platform that can be used to display and analyze performance monitoring data collected from a variety of sources, including RUM tools.
> 
> It's important to note that each type of analysis data has its own advantages and is used in specific contexts. Local data is useful for real-time development and debugging, synthetic data enables performance to be compared and optimized in a reproducible way, while RUM data provides a true view of the user experience. It is generally recommended to use a combination of these different data sources to identify/target problems or opportunities for performance improvement.
> 
> Once you've identified the relevant performance criteria and used the analysis tools to measure your website's current performance, it's essential to stabilize the various criteria before setting improvement targets. Here are the steps to follow:
> 
> 1. Stabilize criteria :
>    - Analyze the results of the performance measurements for each criterion you've identified. Identify areas where improvements are needed, and problems that affect performance stability. For example, LCP depends on the appearance of the widest element on the page. It's important that this element is always the same on the same page type, otherwise there will be no consistent result and optimization will be impossible.
>    - Examine performance metrics over a sufficiently long period of time to detect trends and variations. This will help you understand whether performance problems are constant or sporadic.
>    - Identify factors that could influence performance, such as code changes, server updates, new features, content contribution, etc...
> 
> 2. Setting improvement targets :
>    - Determine the performance levels you want to achieve for each criterion, based on benchmarks, best practices or objectives specific to your application.
>    - Define measurable and realistic objectives for each performance criterion. For example, you could aim to reduce page load time by 20%, improve server response time by less than 200 ms, or reduce bounce rate by 15%.
>    - Prioritize goals according to their impact on your website's user experience and business objectives.
> 
> 3. Plan and implement improvements :
>    - Identify the specific actions to be taken to achieve each improvement objective. This may include code optimizations, server configuration adjustments, resource optimizations, caching, etc.
>    - Establish an action plan for each objective, defining the necessary steps, responsibilities, deadlines and resources required.
>    - Implement improvements iteratively, and measure performance regularly to assess the impact of changes.
>    - Repeat the process of analyzing, stabilizing and improving performance iteratively to continue optimizing your website.
> 
> It's important to bear in mind that website performance is often an ongoing process, as needs, functionalities and conditions change over time. It is therefore advisable to regularly monitor performance and continue to identify new opportunities for improvement to deliver a superior user experience.
> 
> It's important to be careful when planning and implementing performance improvements. Here are some points to consider:
> 
> 1. The performance possibilities are not the same between Single Pages Applications (SPA) and Multiple Pages Applications (MPA).
>     - MPAs are simple to measure: each page is independent. The first page loads resources and then caches certain resources. It's therefore important to be able to distinguish between measurements based on a user's first visit and those based on browsing (which is much faster).
>     - SPAs are more difficult, because apart from the initial page load, there is other page load. Initial loading is often heavier (because of js framework), and therefore slower. Page changes can't be measured simply, which leads to a sharp deterioration in page load criteria, since only the initial load is taken into account. Research is currently underway in browsers to improve the collection of information on page changes in SPAs. This new metric is called soft-navigations and is currently being tested on Chrome and specified by web organizations.
> 2. The performance metrics mentioned above identify a page's performance on paper. They are used by search engines to classify/order search results (seo). However, they are not perfectly representative of user experience. It's possible to have a super-fast LCP or FCP and still have a feeling of slowness or discomfort on the user's side. For example, a page that quickly displays the thumbnail of a film (LCP validation criterion) but takes a long time to display the film's cast list (which is the content the user wants) won't be great for the user experience.
> 
> Sources : 
> - [Measure or die: rapid diagnostics by Jean-Pierre Vincent](https://www.welovespeed.com/2023/interventions/#diagnostics-rapides)
> - [Soft Navigations, it's hard! by Yoav Weiss](https://www.welovespeed.com/2023/interventions/#soft-navigations-are-hard)
> - [How to become a Web Performance Sherlock Holmes by Ludovic Lefebvre](https://www.welovespeed.com/2023/interventions/#devenir-sherlock-holmes)
> - [Reducing RUM noise by Tim Verreecke](https://www.welovespeed.com/2023/interventions/#noise-canceling-rum)
> - [Understanding cognitive bias in performance measurement by Philip Tellis](https://www.welovespeed.com/2023/interventions/#understanding-cognitive-biais-in-performance-measurement)

Reviewer: ```suggestion
When analyzing a website's performance, it's essential to take several key criteria into account. Here are some of the commonly used performance analysis criteria:

![metric.png](/images/posts/2023-06-08-we-love-speed-2023/metric.png)

To measure/evaluate your site's performance according to these criteria, there are 3 types of sources: Local (devtools), Synthetic (CI/CD / online scan tools), RUM (Real User Monitoring).

![metric.png](/images/posts/2023-06-08-we-love-speed-2023/perf-data-source.png)

- Local development tools - devtools :
  - Chrome DevTools: A set of development tools integrated into the Chrome browser, for analyzing performance, debugging JavaScript code, inspecting DOM elements, etc.
  - Firefox DevTools: Tools similar to Chrome DevTools, but specific to the Firefox browser. They offer debugging, performance auditing and inspection of web elements.
  - Lighthouse DevTools (Chrome): A Chrome DevTools extension that provides automated audits to evaluate performance, accessibility, SEO optimization, etc.

- Synthetic testing - CI/CD and online analysis tools :
  - [WebPageTest](https://www.webpagetest.org/): An online tool for evaluating website performance by running tests on different browser configurations, network connections and geographical locations.
  - [Lighthouse CI](https://github.com/GoogleChrome/lighthouse-ci): A Lighthouse-based continuous integration (CI) solution that automates web performance and quality audits with every deployment.

- Real User Monitoring (RUM):
  - [New Relic](https://newrelic.com): An application performance monitoring platform that provides real-time visibility into the performance of your web applications, including usage metrics, load times, etc.
  - [Grafana](https://grafana.com/): A data visualization platform that can be used to display and analyze performance monitoring data collected from a variety of sources, including RUM tools.
  
![logomosaic.png](/images/posts/2023-06-08-we-love-speed-2023/logomosaic.png)

It's important to note that each type of analysis data has its own advantages and is used in specific contexts. Local data is useful for real-time development and debugging, synthetic data enables performance to be compared and optimized in a reproducible way, while RUM data provides a true view of the user experience. It is generally recommended to use a combination of these different data sources to identify/target problems or opportunities for performance improvement.

Once you've identified the relevant performance criteria and used the analysis tools to measure your website's current performance, it's essential to stabilize the various criteria before setting improvement targets. Here are the steps to follow:

1. Stabilize criteria :
  - 🧐 Analyze the results of the performance measurements for each criterion you've identified. Identify areas where improvements are needed, and problems that affect performance stability. For example, LCP depends on the appearance of the widest element on the page. It's important that this element is always the same on the same page type, otherwise there will be no consistent result and optimization will be impossible.
  - 📊 Examine performance metrics over a sufficiently long period of time to detect trends and variations. This will help you understand whether performance problems are constant or sporadic.
  - 💥 Identify factors that could influence performance, such as code changes, server updates, new features, content contribution, etc...

2. Setting improvement targets :
  - 🎯 Determine the performance levels you want to achieve for each criterion, based on benchmarks, best practices or objectives specific to your application.
  - 🎚️ Define measurable and realistic objectives for each performance criterion. For example, you could aim to reduce page load time by 20%, improve server response time by less than 200 ms, or reduce bounce rate by 15%.
  - 🤑 Prioritize goals according to their impact on your website's user experience and business objectives.

3. Plan and implement improvements :
  - 👀 Identify the specific actions to be taken to achieve each improvement objective. This may include code optimizations, server configuration adjustments, resource optimizations, caching, etc.
  - 🗓️ Establish an action plan for each objective, defining the necessary steps, responsibilities, deadlines and resources required.
  - 🚧 Implement improvements iteratively, and measure performance regularly to assess the impact of changes.
  - ♾️ Repeat the process of analyzing, stabilizing and improving performance iteratively to continue optimizing your website.

It's important to bear in mind that website performance is often an ongoing process, as needs, functionalities and conditions change over time. It is therefore advisable to regularly monitor performance and continue to identify new opportunities for improvement to deliver a superior user experience.

It's important to be careful when planning and implementing performance improvements. Here are some points to consider:

1. The performance possibilities are not the same between Single Pages Applications (SPA) and Multiple Pages Applications (MPA).
   - MPAs are simple to measure: each page is independent. The first page loads resources and then caches certain resources. It's therefore important to be able to distinguish between measurements based on a user's first visit and those based on browsing (which is much faster).
   - SPAs are more difficult, because apart from the initial page load, there is other page load. Initial loading is often heavier (because of js framework), and therefore slower. Page changes can't be measured simply, which leads to a sharp deterioration in page load criteria, since only the initial load is taken into account. Research is currently underway in browsers to improve the collection of information on page changes in SPAs. This new metric is called soft-navigations and is currently being tested on Chrome and specified by web organizations.
2. The performance metrics mentioned above identify a page's performance on paper. They are used by search engines to classify/order search results (seo). However, they are not perfectly representative of user experience. It's possible to have a super-fast LCP or FCP and still have a feeling of slowness or discomfort on the user's side. For example, a page that quickly displays the thumbnail of a film (LCP validation criterion) but takes a long time to display the film's cast list (which is the content the user wants) won't be great for the user experience.

Sources :
- [Measure or die: rapid diagnostics by Jean-Pierre Vincent](https://www.welovespeed.com/2023/interventions/#diagnostics-rapides)
- [Soft Navigations, it's hard! by Yoav Weiss](https://www.welovespeed.com/2023/interventions/#soft-navigations-are-hard)
- [How to become a Web Performance Sherlock Holmes by Ludovic Lefebvre](https://www.welovespeed.com/2023/interventions/#devenir-sherlock-holmes)
- [Reducing RUM noise by Tim Verreecke](https://www.welovespeed.com/2023/interventions/#noise-canceling-rum)
- [Understanding cognitive bias in performance measurement by Philip Tellis](https://www.welovespeed.com/2023/interventions/#understanding-cognitive-biais-in-performance-measurement)
```

  - Lighthouse DevTools (Chrome): A Chrome DevTools extension that provides automated audits to evaluate performance, accessibility, SEO optimization, etc.
- Synthetic testing - CI/CD and online analysis tools :
  - [WebPageTest](https://www.webpagetest.org/): An online tool for evaluating website performance by running tests on different browser configurations, network connections and geographical locations.
  - [Lighthouse CI](https://github.com/GoogleChrome/lighthouse-ci): A Lighthouse-based continuous integration (CI) solution that automates web performance and quality audits with every deployment.
- Real User Monitoring (RUM):
  - [New Relic](https://newrelic.com): An application performance monitoring platform that provides real-time visibility into the performance of your web applications, including usage metrics, load times, etc.
  - [Grafana](https://grafana.com/): A data visualization platform that can be used to display and analyze performance monitoring data collected from a variety of sources, including RUM tools.
  
![logomosaic.png](/images/posts/2023-07-01-we-love-speed-2023/logomosaic.png)

> It's important to note that each type of analysis data has its own advantages and is used in specific contexts. Local data is useful for real-time development and debugging, synthetic data enables performance to be compared and optimized in a reproducible way, while RUM data provides a true view of the user experience. It is generally recommended to use a combination of these different data sources to identify/target problems or opportunities for performance improvement.

Reviewer 2: Parler de la conf sur le RUM https://www.youtube.com/watch?v=vfl-C1FZSwE

Sources :
- [Measure or die: rapid diagnostics by Jean-Pierre Vincent](https://www.welovespeed.com/2023/interventions/#diagnostics-rapides)
- [Soft Navigations, it's hard! by Yoav Weiss](https://www.welovespeed.com/2023/interventions/#soft-navigations-are-hard)
- [How to become a Web Performance Sherlock Holmes by Ludovic Lefebvre](https://www.welovespeed.com/2023/interventions/#devenir-sherlock-holmes)
- [Reducing RUM noise by Tim Verreecke](https://www.welovespeed.com/2023/interventions/#noise-canceling-rum)
- [Understanding cognitive bias in performance measurement by Philip Tellis](https://www.welovespeed.com/2023/interventions/#understanding-cognitive-biais-in-performance-measurement)

> ## So, how to actually improve performance ?

Reviewer 2: ```suggestion
## Implement improvement to Web performances
```

> So then, one might ask themselves a question : How do I, actually, improve my performance ? I know what’s the problem, but how do I fix it ?

Reviewer 2: Pas de question a l'utilisateur

Well, obviously, there is no “perfect” answer : What might be enough for one website might not suffice for another, and some websites might also handle their own code well enough so that some improvements are not needed.
No matter what, however, here are some clues as to how you could make some improvements ! 

### Use bfcache

Did you know that around 10% of a user's navigation consists of clicking on the “Go back” button of their browser ? More often than not, however, when doing that, our users need to wait for the page to load again, despite the fact that the browser already did that work for us !

> Browser developers already noticed that, however, and we recently began to see the apparition of something named the Back/Forward cache (Also simply named bfcache), who’s goal is to do exactly that : Cache the page the user is leaving, so that if they go back using their browser features, they’ll see the cached page instead. You can find more about [this amazing cache here](https://web.dev/bfcache/) !

Reviewer 2: Mettre lien vers le talk : https://www.youtube.com/watch?v=fJrKeh-Np8c (Probablement timestamp)

### Preloading, Preconnecting and HTML

> There is one important rule that you need to always remember when doing HTML : It is parsed line by line, and it can be paused. When your browser is trying to render your page, if it runs across something trying to call an url (Ex: loading a css file), then it’ll wait until said css file is fully loaded, and then again it’ll keep going !

Reviewer 2: Introduire avec le même talk qu'au-dessus

However, you can use the HTML property `rel` to prepare some of that stuff ahead. For example, `rel=”preload”` will tell the browser to start loading said css asynchronously, as soon as they arrive on the page ! Using an API ? Then you can use `rel=”preconnect”` to start your handshakes as soon as you’re on your page, so that your API is fully ready to send you packages as soon as you need it !

For more details about all of that, feel free to check [Mozilla’s documentation](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/rel) , it contains even more attribute values than the two examples here.
If you want your resource to be even more high-priority, you can go the extra step, and [use the `fetchpriority` attribute](https://web.dev/fetch-priority/), which will be more than glad to fetch as soon as possible your corresponding resources ! As easy as it sounds, [it can have immense benefit](https://twitter.com/senthil_hi/status/1589676464215326720).

> However, please be mindful about using these ! Your browser can only do so much at the same time : And if you’re overusing these, then for your browser everything will become a top-priority, and it won’t be able to load specifically what you need the most. As a result, you’ll go back to your first issue. Always remember that if everything is important, then nothing is !

Reviewer 2: Un peu d'espacement

> ### Lazy Loading

Reviewer 2: ```suggestion
### Lazy...
```

> Images, and especially high-quality ones, can take up a lot of your time to load. If you’re using a lot of images on your website, you’ve probably noticed it by now. However, quite often, you’re not always displaying every image to your user at the same time : Maybe some are located down the page, maybe some are only seen by an user after some time or if they do a specific action : The point is, you don’t need to load all of these images right from the get-go. That’s what lazy-loading is for : Only loading the images the user sees right now, and then only, if we have some time, start to load what the user isn’t seeing yet.

Reviewer 2: Pareil, même conf

 Here again, you have an HTML attribute that can take care of everything for you : Named `loading`, you can pass the value `lazy` to it and it’ll only load your image when approaching the screen. Find out more about that attribute [over here](https://web.dev/browser-level-image-lazy-loading/) !

>  Do note, however, that lazy-loading is not limited to images only : Maybe you could lazyload your components ? Your thirdparties ? Do you really need your chat component to be ready and available if the user did not click on the chat icone yet ? Do you need the youtube thirdparty to be ready to play the video, if the user still hasn't clicked on said video ? Maybe sometimes, it’s for the best to increase an action response time if it means that your website is loaded faster. 

Reviewer 2: Les sauts de ligne c'est pas que a l'élastique

>  Do note, however, that lazy-loading is not limited to images only : Maybe you could lazyload your components ? Your thirdparties ? Do you really need your chat component to be ready and available if the user did not click on the chat icone yet ? Do you need the youtube thirdparty to be ready to play the video, if the user still hasn't clicked on said video ? Maybe sometimes, it’s for the best to increase an action response time if it means that your website is loaded faster. 

Reviewer 2: Lazy hydrate ici

> ### 103 Early hints

Reviewer 2: Devrait être une des dernières propositions

> Do note that this is an experimental feature ⚠️

> When downloading files from a server (Such as font files, css files, or others), at first, the server does not know what size of packages your browser is able to handle. It will thus need to “warm up” first, sending more and more packages at once until the browser has reached its limit, at which point it will adapt and keep sending at the highest possible (and manageable) speed.

Reviewer 2: Citer la conf

The 103 Early hints feature allows your browser, while it is connected to the server and not downloading anything (Such as when using a preconnect), to start that warmup round instead of being idle : That way, when the server will be needed to download, it’ll be able to immediately start at maximum output. This will increase the speed at which your application is able to download content from your server, therefore decreasing waiting time for your end user. 

### Take care of your CMP !

> Nowadays, a lot of websites (And of course, us too), are using a lot of third parties. Whether it is in order to monitor what your user is doing so that you can track their issues more precisely, to gather some data for your teams or to send information for you to sell, there are more and more thirdparties through the internet.

Reviewer 2: Lien vers conf : https://www.youtube.com/watch?v=q3M4he-QxCA

However, each new thirdparty comes with its own compounding issues : Namely, each is adding another domain, meaning another need to connect to a distant API, reheat the connection… And all of that is taking network time, precious time that you actually need to load a lot more data for your website ! And we’re not even speaking of them using custom SDKs, using your threads and computing time to run their javascript…
First of all, it is important to understand how your thirdparties work : You CANNOT, and should NEVER try to preconnect to them. Preconnecting to a thirdparty discloses your user IP, which according to GDPR laws at the time of writing, is considered a personnel data. If your user does not consent to sharing their data, you should never try to connect in the first place.

> Therefore, you need to collect the user's consent to be able to connect to your thirdparties API : But to collect said consent, you need your Consent Manager to be booted and ready. Therefore, the first improvement to make is to optimize your CMP : Make sure it loads fast, make sure it shares the user’s consent with you as soon as possible so that you’re ready to go fetch those thirdparties as soon as you can !

Reviewer 2: El pavé

### Self-host blocking thirdparties

> Just like we mentioned, many thirdparties are vital : So much that sometimes, you also need them to load in order to display something at all. Maybe, for example, it’s the thirdparty responsible for your A/B testing : Of course you need it to know what your user will see !

Reviewer 2: Phrases courtes, citer la conf d'au-dessus

However, it is also important to remember that since your browser has limited network capabilities, it will try to prioritize its requests : And requests that are going to another domain are far down the line. 
If you have blocking thirdparties that are mandatory, moving them to your own domain is a good idea to increase your load time : Your browser will prioritize more these requests, your website will know what to display sooner, and will display the right content as soon as possible !

### Use Server-side Rendering (SSR) !

> Easier said than done, Server-Side Rendering is a wonderful tool to reduce what’s called your “Total Blocking Time” in your application. By allowing your code to be pre-rendered in your server before being sent to your browser, you can dramatically decrease the time it takes for an user to be able to interact with your page. 

Reviewer 2: Retrouver la conf

> Easier said than done, Server-Side Rendering is a wonderful tool to reduce what’s called your “Total Blocking Time” in your application. By allowing your code to be pre-rendered in your server before being sent to your browser, you can dramatically decrease the time it takes for an user to be able to interact with your page. 

Reviewer 2: ```suggestion
Easier said than done, Server-Side Rendering is a wonderful tool to reduce what’s called your “Total Blocking Time” in your application. By allowing your code to be pre-rendered in your server before being sent to your browser, you can drastically decrease the time it takes for an user to be able to interact with your page. 
```

You can then fine tune your SSR using librairies and packages such as the [Vue Lazy Hydration](https://www.npmjs.com/package/vue-lazy-hydration) one. 
However, this is not a perfect solution : It requires a lot of time, comprises a lot of technically difficult topics, and may require you to change your framework altogether. But there’s a silver lining : At the time of writing, more and more frameworks are trying to pick up and make more tools available to us to easily access SSR : Astro or Nuxt for example.

> All in all, performance is definitely anything but an easy task to do. It is not a magic formula that will make all your problems go away : And even if you manage to hold a long workshop of several weeks (Or even months !) that will finally fix your performance issues, if it does not become a habit in your developer teams, it will never be over. Performance is a habit, built through practice and good communication : But it can be achieved.

Reviewer 2: Pas d'espace avant la ponctuation

---
layout: post
title: "Bedrock at We Love Speed 2023"
description: We tell you all about web performance in 2023!
author: team_frontend
tags: [conference, paris, tech, webperf]
color: rgb(251,87,66)
thumbnail: /images/posts/we-love-speed-2023/thumbnail.jpeg
---

The Frontend Bedrock team was present at the 2023 edition of the _We Love Speed_ conference in Paris on May 10. Its aim is to share as widely as possible knowledge and experience in the field of web performance.

Webperf, or web performance, refers to the set of techniques and best practices aimed at optimizing the loading speed and user experience of websites. It covers various aspects such as page loading time, interface responsiveness, animation fluidity, server request processing capacity, etc.

> Webbperf is a major challenge in today's web industry, influencing user experience, website visibility and overall commercial success. This article summarizes the topics discussed at this year's _We Love Speed_ conference and gives an overview of the 2023 approach to web performance.

Reviewer 3: ```suggestion
Webperf is a major challenge in today's web industry, influencing user experience, website visibility and overall commercial success. This article summarizes the topics discussed at this year's _We Love Speed_ conference and gives an overview of the 2023 approach to web performance.
```

> ## Why optimize your sites' webperf?

Reviewer 3: ```suggestion
## Why optimize your site's webperf?
```
Petit doute là dessus ?

Author: Deepl dit que pour parler de "sites" au pluriel, ça se fait 🙂

Web performance is essential, as it has a direct impact on user satisfaction, engagement and the overall success of a website. Fast performance enables users to access information quickly, interact without delay and navigate smoothly. On the other hand, slow loading times, frozen interfaces or high response times can lead to user frustration, lower engagement and higher bounce rates.

A lot of articles on the net have proved **the impact of the webperf on users' engagement** (here is [an great article from Google](https://web.dev/why-speed-matters/) or [hundreds of case studies](https://wpostats.com/) from WPO stats). So, depending on the site's field, it's relatively easy to link the webperf approach to an objective of user views or loyalty, but also with the brand image. And the company's governance will be able to quickly measure the benefits in terms of return on investment.

### What about SEO?

> Moreover, search engines such as Google also place give a relative importance on website performance in their ranking algorithms. Faster, more optimized sites often benefit from better positioning in search results, which can have an impact on their visibility and traffic.

Reviewer 3: ```suggestion
Moreover, search engines such as Google also give a relative importance on website performance in their ranking algorithms. More optimized sites often benefit from better positioning in search results, which can have an impact on their visibility and traffic.
```

However, [Philippe Yonnet's talk _"Quel est le véritable impact des problèmes de web performance sur le SEO ?"_](https://www.youtube.com/watch?v=uWqQCOX52i4) 🇫🇷 reveals that actually the Google webperf scoring is in reality quite low in all the parameters taken into account in the ranking algorithm. Relevant content is always the first factor in top positioning. As was the case with responsive design a few years ago, the Google's communication around webperf is aimed more at **encouraging good practices** in the web community than at really penalizing slow sites.

## How do you take up a performance project?

### Build a long-term strategy based on usages

You need a strategy and a speech! [Boris Schapira](https://www.youtube.com/watch?v=wh0zCqZRuSs) 🇫🇷 defines 6 steps:

1. Identify the sticking point
2. Associate a value to the project for the business
3. Invest time for the optimizations
4. Evaluate the ROI

> 5. Sustaine the approach over time (it should never be a one shot project)

Reviewer 3: ```suggestion
5. Sustain the approach over time (it should never be a one shot project)
```

- take a period of time where the measure is representative (e.g. on our streaming sites, the activity period is mainly in the evening)
- take into account the fact that different actions or events on your site may not involve the same users (so not the same devices, etc.)
- take "short" indicators to measure the impact of the performance (e.g. if users are on a product list, monitor the viewing of the product page and not the purchase of the product)
- take into account the market you're addressing (e.g. food click & collect users are more likely to wait the loading of the next page than users on a retail site)

### Take cognitive biases into account to prioritise optimisations

When it comes to prioritizing performance subjects, there are several strategies to consider. [Philip Tellis in his talk _"Understanding Cognitive Biases in Performance Measurement"_](https://www.youtube.com/watch?v=6KwW5v5rce4) 🇬🇧 shed light on the psychological aspect of performance, focusing on cognitive biases.

> Tellis made a powerful statement: _"If you have a brain, you have a bias_". Let's explore three examples that demonstrate how biases can impact your work and how you can leverage them to your advantage.

Reviewer 3: ```suggestion
Philip made a powerful statement: _"If you have a brain, you have a bias_". Let's explore three examples that demonstrate how biases can impact your work and how you can leverage them to your advantage.
```

#### Slowest is the norm

The negativity bias reveals that negative experiences tend to leave a stronger impression than positive ones, even if their intensity is the same. This is crucial to keep in mind when optimizing your app's performance. Ideally, the slowest page should not be more than 15 times slower than the fastest page. Users might perceive the slower page as the norm for your app.

To counteract this bias, it's important to be transparent with your users. You can display a message indicating that the page is slow and assure them that efforts are underway to resolve the issue. Interestingly, slow pages have been associated with a 38% increase in heart rate, similar to watching a horror movie. **By prioritizing performance and being honest with your users, you can mitigate stress and anxiety.**

> #### Take care at the beginning and end

Reviewer 3: Which bias it was ?

Author: The serial position effect, it's mentioned at the beginning of the paragraph

2. Setting improvement targets :
- **Determine the performance levels** you want to achieve for each criterion, based on benchmarks, best practices or objectives specific to your application.
- **Define measurable and realistic objectives** for each performance criterion. For example, you could aim to reduce page load time by 20%, improve server response time by less than 200 ms, or reduce bounce rate by 15%.
- **Prioritize goals** according to their impact on your website's user experience and business objectives.

3. Plan and implement improvements :
- **Identify the specific actions to be taken** to achieve each improvement objective. This may include code optimizations, server configuration adjustments, resource optimizations, caching, etc.
- **Establish an action plan** for each objective, defining the necessary steps, responsibilities, deadlines and resources required.
- **Implement improvements** iteratively, and measure performance regularly to assess the impact of changes.
- **Repeat the process** of analyzing, stabilizing and improving performance iteratively to continue optimizing your website.

> t's important to bear in mind that website performance is often an ongoing process, as needs, functionalities and conditions change over time. It is therefore advisable to regularly monitor performance and continue to identify new opportunities for improvement to deliver a superior user experience.

Reviewer 3: ```suggestion
It's important to bear in mind that website performance is often an ongoing process, as needs, functionalities and conditions change over time. It is therefore advisable to regularly monitor performance and continue to identify new opportunities for improvement to deliver a superior user experience.
```

It is also important to be careful when planning and implementing performance improvements. Here are some points to consider:

1. The performance possibilities are not the same between **Single Pages Applications (SPA)** and **Multiple Pages Applications (MPA)**.
   - MPAs are simple to measure: each page is independent. The first page loads resources and then caches certain resources. It's therefore important to be able to distinguish between measurements based on a user's first visit and those based on browsing (which is much faster).

>    - SPAs are more difficult, because apart from the initial page load, there is other page load. Initial loading is often heavier (because of js framework), and therefore slower. Page changes can't be measured simply, which leads to a sharp deterioration in page load criteria, since only the initial load is taken into account. Research is currently underway in browsers to improve the collection of information on page changes in SPAs. This new metric is called soft-navigations and is currently being tested on Chrome and specified by web organizations. Yoav Weiss explains this in his talk ["Soft Navigations are hard!!"](https://www.youtube.com/watch?v=pQLRoABCnqg) 🇬🇧.

Reviewer 3: ```suggestion
   - SPAs are more difficult, because apart from the initial page load, there is other page load. Initial loading is often heavier (because of js framework), and therefore slower. Page changes can't be measured simply, which leads to a sharp deterioration in page load criteria, since only the initial load is taken into account. Research is currently underway in browsers to improve the collection of information on page changes in SPAs. This new metric is called `soft-navigation` and is currently being tested on Chrome and specified by web organizations. Yoav Weiss explains this in his talk ["Soft Navigations are hard!!"](https://www.youtube.com/watch?v=pQLRoABCnqg) 🇬🇧.
```

> - Do users quickly see content they can trust? ➡️ it's measured though the _Largest Contentful Paint_ (LCP)
> - Can users quickly interact with the page in a qualitative way? ➡️ it's measured though the _First Input Delay_ (FID)

But these KPI are a limited reflection of user experience, and it is hard to link as it to a representative value. **You need to correlate these measures with usages**. There is no magic solution for this but segmenting sessions is a good practice to focus on relevant optimizations:

- take a period of time where the measure is representative (e.g. on our streaming sites, the activity period is mainly in the evening)
- take into account the fact that different actions or events on your site may not involve the same users (so not the same devices, etc.)
- take "short" indicators to measure the impact of the performance (e.g. if users are on a product list, monitor the viewing of the product page and not the purchase of the product)
- take into account the market you're addressing (e.g. food click & collect users are more likely to wait the loading of the next page than users on a retail site)

### Priorization

### Community

Maxime ?

> ## Avec quels moyens s’attaquer à la performance ?

Reviewer: ```suggestion
## What are the means to target performance issues?
```
